File: cpp/fory/util/buffer_test.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <iostream>
#include <limits>
#include <utility>
#include <vector>

#include "gtest/gtest.h"

#include "fory/util/buffer.h"

namespace fory {

TEST(Buffer, to_string) {
  std::shared_ptr<Buffer> buffer;
  allocate_buffer(16, &buffer);
  for (int i = 0; i < 16; ++i) {
    buffer->unsafe_put_byte<int8_t>(i, static_cast<int8_t>('a' + i));
  }
  EXPECT_EQ(buffer->to_string(), "abcdefghijklmnop");

  float f = 1.11;
  buffer->unsafe_put<float>(0, f);
  EXPECT_EQ(buffer->get<float>(0), f);
}

void check_var_uint32(int32_t start_offset, std::shared_ptr<Buffer> buffer,
                      int32_t value, uint32_t bytes_written) {
  uint32_t actual_bytes_written = buffer->put_var_uint32(start_offset, value);
  EXPECT_EQ(actual_bytes_written, bytes_written);
  uint32_t read_bytes_length;
  int32_t var_int = buffer->get_var_uint32(start_offset, &read_bytes_length);
  EXPECT_EQ(value, var_int);
  EXPECT_EQ(read_bytes_length, bytes_written);
}

TEST(Buffer, TestVarUint) {
  std::shared_ptr<Buffer> buffer;
  allocate_buffer(64, &buffer);
  for (int i = 0; i < 32; ++i) {
    check_var_uint32(i, buffer, 1, 1);
    check_var_uint32(i, buffer, 1 << 6, 1);
    check_var_uint32(i, buffer, 1 << 7, 2);
    check_var_uint32(i, buffer, 1 << 13, 2);
    check_var_uint32(i, buffer, 1 << 14, 3);
    check_var_uint32(i, buffer, 1 << 20, 3);
    check_var_uint32(i, buffer, 1 << 21, 4);
    check_var_uint32(i, buffer, 1 << 27, 4);
    check_var_uint32(i, buffer, 1 << 28, 5);
    check_var_uint32(i, buffer, 1 << 30, 5);
  }
}

void check_var_uint64(int32_t start_offset, std::shared_ptr<Buffer> buffer,
                      uint64_t value, uint32_t bytes_written) {
  uint32_t actual_bytes_written = buffer->put_var_uint64(start_offset, value);
  EXPECT_EQ(actual_bytes_written, bytes_written);
  uint32_t read_bytes_length;
  uint64_t var_int = buffer->get_var_uint64(start_offset, &read_bytes_length);
  EXPECT_EQ(value, var_int);
  EXPECT_EQ(read_bytes_length, bytes_written);
}

TEST(Buffer, TestVarUint64) {
  std::shared_ptr<Buffer> buffer;
  allocate_buffer(256, &buffer);
  const std::vector<std::pair<uint64_t, uint32_t>> cases = {
      {0, 1},
      {1, 1},
      {127, 1},
      {128, 2},
      {16383, 2},
      {16384, 3},
      {2097151, 3},
      {2097152, 4},
      {268435455, 4},
      {268435456, 5},
      {34359738367ULL, 5},
      {34359738368ULL, 6},
      {4398046511103ULL, 6},
      {4398046511104ULL, 7},
      {562949953421311ULL, 7},
      {562949953421312ULL, 8},
      {72057594037927935ULL, 8},
      {72057594037927936ULL, 9},
      {std::numeric_limits<uint64_t>::max(), 9},
  };
  for (int i = 0; i < 32; ++i) {
    for (const auto &entry : cases) {
      check_var_uint64(i, buffer, entry.first, entry.second);
    }
  }
}

TEST(Buffer, TestGetBytesAsInt64) {
  std::shared_ptr<Buffer> buffer;
  allocate_buffer(64, &buffer);
  buffer->unsafe_put<int32_t>(0, 100);
  int64_t result = -1;
  EXPECT_TRUE(buffer->get_bytes_as_int64(0, 0, &result).ok());
  EXPECT_EQ(result, 0);
  EXPECT_TRUE(buffer->get_bytes_as_int64(0, 1, &result).ok());
  EXPECT_EQ(result, 100);
}
} // namespace fory

int main(int argc, char **argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: go/fory/buffer_test.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package fory

import (
	"testing"

	"github.com/stretchr/testify/require"
)

func TestVarint(t *testing.T) {
	err := &Error{}
	for i := 1; i <= 32; i++ {
		buf := NewByteBuffer(nil)
		for j := 0; j < i; j++ {
			buf.WriteByte_(1) // make address unaligned.
			buf.ReadByte(err)
		}
		// Zigzag encoding doubles positive values: zigzag(n) = n * 2 for positive n
		// So boundary values for positive numbers are:
		// 1 byte: 0-63 (zigzag 0-126, fits in 7 bits)
		// 2 bytes: 64-8191 (zigzag 128-16382, fits in 14 bits)
		// 3 bytes: 8192-1048575 (zigzag fits in 21 bits)
		// 4 bytes: 1048576-134217727 (zigzag fits in 28 bits)
		// 5 bytes: 134217728+ (zigzag fits in 32 bits)
		checkVarint(t, buf, 1, 1)
		checkVarint(t, buf, 63, 1)        // max 1-byte positive: zigzag(63)=126
		checkVarint(t, buf, 64, 2)        // min 2-byte positive: zigzag(64)=128
		checkVarint(t, buf, 8191, 2)      // max 2-byte positive: zigzag(8191)=16382
		checkVarint(t, buf, 8192, 3)      // min 3-byte positive: zigzag(8192)=16384
		checkVarint(t, buf, 1048575, 3)   // max 3-byte positive
		checkVarint(t, buf, 1048576, 4)   // min 4-byte positive
		checkVarint(t, buf, 134217727, 4) // max 4-byte positive
		checkVarint(t, buf, 134217728, 5) // min 5-byte positive
		checkVarint(t, buf, MaxInt32, 5)
		checkVarintWrite(t, buf, -1)
		checkVarintWrite(t, buf, -1<<6)
		checkVarintWrite(t, buf, -1<<7)
		checkVarintWrite(t, buf, -1<<13)
		checkVarintWrite(t, buf, -1<<14)
		checkVarintWrite(t, buf, -1<<20)
		checkVarintWrite(t, buf, -1<<21)
		checkVarintWrite(t, buf, -1<<27)
		checkVarintWrite(t, buf, -1<<28)
		checkVarintWrite(t, buf, MinInt8)
		checkVarintWrite(t, buf, MinInt16)
		checkVarintWrite(t, buf, MinInt32)
	}
}

func checkVarint(t *testing.T, buf *ByteBuffer, value int32, bytesWritten int8) {
	err := &Error{}
	require.Equal(t, buf.WriterIndex(), buf.ReaderIndex())
	actualBytesWritten := buf.WriteVarint32(value)
	require.Equal(t, bytesWritten, actualBytesWritten)
	varInt := buf.ReadVarint32(err)
	require.Equal(t, buf.ReaderIndex(), buf.WriterIndex())
	require.Equal(t, value, varInt)
}

func checkVarintWrite(t *testing.T, buf *ByteBuffer, value int32) {
	err := &Error{}
	require.Equal(t, buf.WriterIndex(), buf.ReaderIndex())
	buf.WriteVarint32(value)
	varInt := buf.ReadVarint32(err)
	require.Equal(t, buf.ReaderIndex(), buf.WriterIndex())
	require.Equal(t, value, varInt)
}
